### PR TITLE
[agent-e][issue #2360] Own project-local dev scratch epochs

### LINKED ISSUE #2360: [P1][serve][dev] Own project-local session-durable SQLite scratch epochs
## Parent and approved boundary

Parent: #2276
Depends on: #2359
Gate E: https://github.com/division-sh/swarm/issues/2276#issuecomment-5430144678
Approved audit Revision 2: https://github.com/division-sh/swarm/issues/2276#issuecomment-5430046055

This is Child B of the approved three-child #2276 sequence.

## Chosen working failure class

`project_local_dev_scratch_coordinate_epoch_exclusion_reclamation_replacement_and_source_ownership`

## Product posture

- Exactly one `swarm serve --dev` runtime per canonical project, independent of `--context` and `--swarm-dir`.
- `--context` labels endpoint discovery only and never selects/partitions a store.
- Project-local contracts use `<canonical-project>/.swarm/stores/dev-scratch.db`; normal durable `dev.db` is untouched.
- Borrowed-contract dev, authored/custom SQLite paths, and effective/explicit PostgreSQL fail closed with teaching remediation.
- One process loads one source generation. No hot reload.
- The epoch remains on disk after shutdown for diagnosis and is replaced as a whole before the next admitted dev invocation.

## Canonical owner

One typed `internal/store/devscratch.EpochAuthority` owns:

`{canonical_project_root, canonical_scratch_coordinate, epoch_id, retained_exclusive_lock}`

The owner:

1. derives the exact canonical project-local coordinate without context/PID/Swarm-dir authority;
2. obtains a retained nonblocking kernel lock on a stable adjacent authority file before deletion;
3. treats lock ownership, not file existence/PID/descriptor state, as live-versus-stale proof;
4. removes the DB and `-wal`/`-shm`/`-journal` only while holding the token and fails before open on any cleanup error;
5. layers into existing SQLite construction/file possession while retaining the epoch lock;
6. releases only after runtime join and selected-store close;
7. is automatically reclaimable after process crash;
8. fails closed on ambiguous aliases/filesystems.

Current context-name/Swarm-dir-scoped `AcquireProjectClaim` is invalid scratch authority and must be removed from the dev scratch path, not wrapped or adopted.

## Manifestations and proof

The PR must cover every SCR-1..SCR-16 row in Audit Revision 2, including real helper-process proof for live contention, crash before open, and crash after open. It must also prove:

- two explicit context names collide at the same project owner;
- different `--swarm-dir` values collide at the same project owner;
- canonical/symlink-equivalent inputs converge;
- borrowed/custom/PostgreSQL postures mutate nothing;
- partial sidecar cleanup cannot open the store;
- the epoch token cannot release before store close;
- next invocation sees no predecessor facts;
- canonical private serve ingest persists the exact bundle/source fact;
- retired Builder project reload/source replacement is removed.

## Governing spec

Update in the same PR:

- `platform-spec.yaml#engine.runtime_store_backend_selection`
- `platform-spec.yaml#engine.runtime_core_persistence_store_contracts.selected_store_composition.required_serve_bundle_ingest`
- `platform-spec.yaml#multi_bundle_persistence.persistence_model.serve_ingest_projection`
- `platform-spec.yaml#cli_specification.foundations.local_runtime_state_authority`
- context registration/ownership and retired Builder transport clauses

No migration, row cleanup, legacy claim adoption, PID/file-existence authority, per-context store, compatibility source, or memory fallback.

## Watchlist and splits

Canonical watchlist: `swarm-docs@18a2eee`, node `runtime_store_backend_default_and_sqlite_portability`.

DB-loaded SQLite `--bundle-hash` remains explicit unsupported/watchlist-only. Closed #1209/PR #1211 is historical parked-Dynamic-Load authority; open #2277 owns changed-bundle fork-context design.

## Closure

This PR eliminates the complete Child B class. #2276 remains open; final same-invocation fork composition is execution-proven by Child C.

## tracker
```yaml
parent: 2276
blockers: [2359]
score: 70
sequence: child_b
closes_parent: false
```

### PR DESCRIPTION
**Pull request by yazzaoui**
## Summary

- add one crash-reclaimable `internal/store/devscratch.EpochAuthority` keyed by the canonical project coordinate
- make `swarm serve --dev` use a dedicated fresh SQLite scratch epoch while leaving normal `dev.db` untouched
- persist the exact dev bundle/source through canonical private serve ingest and remove reachable project reload/source replacement
- reject borrowed contracts, authored SQLite paths, PostgreSQL, aliases, unsupported filesystems, and concurrent owners before store mutation

## Governing authority

Binding sections updated in this PR:

- `platform-spec.yaml#engine.runtime_store_backend_selection`
- `platform-spec.yaml#engine.runtime_core_persistence_store_contracts.selected_store_composition.required_serve_bundle_ingest`
- `platform-spec.yaml#multi_bundle_persistence.persistence_model.serve_ingest_projection`
- `platform-spec.yaml#cli_specification.foundations.local_runtime_state_authority`
- context registration/ownership and retired Builder source-replacement clauses

The approved boundary and Gate E are recorded on #2276 and #2360.

## Semantic migration

- **New canonical owner:** `internal/store/devscratch.EpochAuthority` owns `{canonical_project_root, canonical_scratch_coordinate, epoch_id, retained_exclusive_lock}`. `SelectedStoreLifecycle` retains that epoch through runtime join and successful selected-store close.
- **Old paths now invalid:** dev ephemeral source construction, private-ingest bypass, context-name/Swarm-dir `AcquireProjectClaim` as scratch authority, custom/shared store adoption, and the production test reload hook.
- **Production paths checked:** CLI/local-state resolution, context reconciliation, SQLite selected construction, private serve ingest, runtime source planning, shutdown/store close, next invocation, public health/run readback, and source replacement reachability.
- **Migration complete:** yes for `project_local_dev_scratch_coordinate_epoch_exclusion_reclamation_replacement_and_source_ownership`. No migration, compatibility, PID/file-existence liveness, row cleanup, per-context store, memory fallback, or hot reload remains.

## Proof

- `go run ./cmd/swarm-test`
- `go test ./internal/store/devscratch -count=3`
- focused CLI, context-registration, persisted-ingest, runtime scratch, no-reload, conformance, release-binary, authority-registry, and supported-mode matrices

SCR-15 same-invocation SQLite fork composition remains explicitly tracked by #2361.

Closes #2360
Part of #2276

### CODE REVIEW

**Comment by yazzaoui** [comment]
## Post-Implementation Proof Audit

### Semantic boundary and closure

- **Exact concepts changed:** canonical project-local dev scratch coordinate selection; crash-reclaimable epoch exclusion and replacement; descriptor reconciliation under an opaque epoch grant; selected-store/epoch shutdown ordering; exact persisted dev bundle/source ingest; one-source-generation/no-reload serve posture.
- **Chosen working failure class:** `project_local_dev_scratch_coordinate_epoch_exclusion_reclamation_replacement_and_source_ownership`.
- **Immediate parent class:** `local_runtime_source_and_preopen_store_lifetime_ownership`.
- **Broader parent class:** `sqlite_selected_public_bundle_and_fork_capability_parity_with_session_scratch_source_ownership` in #2276, under selected-store backend parity/source-lifecycle conformance in #2275.
- **Issue shape:** #2360 is class-shaped, not symptom-shaped. The old ephemeral source branch, context claim, custom path acceptance, and restart behavior were entry points rather than the audit boundary.
- **Achieved closure level:** complete Child B failure-class elimination. The selected project-local dev posture has one coordinate, one liveness authority, one replacement owner, one persisted source interpretation, and one ordered release path.
- **Parent state:** #2276 remains open. Its remaining child tail is #2361, complete selected-contract SQLite `run.fork` parity plus same-invocation composition. Estimate: one bounded child PR, high confidence; implementation did not change that estimate.

### Canonical owners after the change

| Concept | Canonical owner | Why it is semantic authority |
|---|---|---|
| Supported dev store posture | `internal/cliapp.ResolveDevScratch` | Purely admits project-local/default-SQLite inputs and derives the only allowed scratch selection before mutation. |
| Coordinate, epoch, and replacement authority | `internal/store/devscratch.EpochAuthority` | Owns `{canonical_project_root, canonical_scratch_coordinate, epoch_id, retained_exclusive_lock}` and is the only operation able to prepare predecessor replacement. |
| Opened store plus epoch release order | `internal/store/devscratch.SelectedStoreLifecycle` | Makes epoch release unavailable until selected-store close succeeds; activated close still requires the process join receipt. |
| Exact boot source projection | `internal/runtime/contracts.BuildBundleCatalogProjection` plus selected `internal/bundlecatalog.ServeIngestWriter` | Produces and commits the authoritative catalog record on both selected stores. |
| Runtime source fact | `internal/serveapp.prepareServeBundleSource` and `internal/runtime/correlation.BundleSourceFact` | There is one persisted-only planner; the dev boolean and ephemeral success branch are gone. |
| Endpoint discovery metadata | `internal/cliapp.ServeProjectContextRegistration` | Consumes a project-bound opaque epoch grant but owns no store exclusion or release authority. |

These are real semantic owners. None is merely the first local file encountered.

### Touched consumers and systematic-consumption audit

| Consumer/seam | Final classification | Exact accounting |
|---|---|---|
| `swarm serve --dev` CLI/config input | moved to canonical owner | `ResolveLocalRuntimeState` resolves provenance, then `ResolveDevScratch` admits or rejects before filesystem mutation. |
| Backend/path selection | moved to canonical owner | Selected SQLite receives `SourceProjectScratch`; context name and Swarm-dir are absent from coordinate derivation. |
| Same-project context descriptor reconciliation | moved to canonical owner | Requires `RegistrationGrant`; same-project descriptors are replaceable metadata, and cross-project global name collisions still fail closed. |
| Context/Swarm-dir `AcquireProjectClaim` | different concept | Removed from dev scratch. It remains only for descriptorless local foreground `swarm run start`, where it guards a different startup posture. |
| Predecessor DB/WAL/SHM/journal replacement | moved to canonical owner | All artifacts are validated before deletion and replacement requires the retained epoch lock. Any error prevents selected-store open. |
| SQLite selected construction/file possession | consumes canonical owner | Opens only the prepared exact coordinate, then is bound into `SelectedStoreLifecycle`; post-open possession does not authorize predecessor deletion. |
| Runtime activation and process work | consumes canonical owner | Existing selected activation is retained; activated close requires the exact process join receipt before epoch release. |
| Unactivated startup failures | consumes canonical owner | Store close must succeed before lock release; close failure keeps the coordinate contended. |
| Private serve ingest, PostgreSQL/SQLite | moved to canonical owner | Every contracts-source plan requires `ServeIngestWriter`; no loader- or dev-specific persistence bypass remains. |
| Run/source stamping | consumes canonical owner | New dev runs carry the exact catalog hash and `bundle_source=persisted`. |
| Next dev invocation | consumes canonical owner | Kernel lock reclamation authorizes full predecessor replacement before open; predecessor bundle/run facts are absent. |
| Runtime identity, health, `swarm test`, and public run readback | execution-proven consumer | Observe the retained persisted source; changed disk input cannot replace the running generation. |
| Builder project reload test hook | invalid and removed | `ServeOptions.TestRuntimeProjectReloadHook` and its production wiring are deleted. Dev construction explicitly disables source replacement. |
| Internal replacement machinery | non-live for this class | No production dev/API/test-hook caller can invoke `ReloadProject`; close/quiescence use separate lifecycle operations. |
| Non-dev selected SQLite/PostgreSQL | different semantic concept | Keeps normal durable store behavior and canonical persisted serve ingest. |
| Borrowed contracts, custom SQLite, PostgreSQL dev | fail-closed sibling manifestations | Rejected before authority acquisition/store mutation; no compatibility or alternate namespace exists. |
| DB-loaded SQLite `--bundle-hash` | explicit split | Remains unsupported/watchlist-only under the approved #2276 ruling. |
| Same-invocation selected-contract SQLite fork | explicit split | SCR-15 is #2361 and closes the final parent composition after this PR merges. |

The production census covered `internal/cliapp`, `internal/serveapp`, `internal/store/backendselection`, `internal/store/selected`, private serve ingest, runtime source planning, context registration, shutdown, conformance ledgers, and release E2E. No live same-concept interpreter remains.

### Old paths now invalid

- Dev `bundle_source=ephemeral` construction and no-row ingest bypass.
- The source planner's dev boolean, ignored compatibility parameter, and ephemeral success path.
- Context-name/Swarm-dir-scoped `AcquireProjectClaim` as scratch liveness or deletion authority.
- Descriptor, PID, network probe, file existence, or DB contents as live-versus-stale proof.
- Normal `dev.db`, configured/custom SQLite, borrowed-project storage, PostgreSQL, per-context DBs, memory fallback, or legacy adoption as dev scratch targets.
- Public/test-hook source replacement during a dev epoch.
- No-op claim-era registration release compatibility surface.

No old path survives as an authoritative dev scratch producer, reader, or interpreter.

### Manifestation coverage

| ID | Manifestation | Disposition | Exact proof |
|---|---|---|---|
| SCR-1 | same project/default context collision | reproduced and fixed | `TestDevScratchEpochDefaultContextOwnsCanonicalCoordinate` |
| SCR-2 | same project/two explicit contexts | reproduced and fixed | `TestDevScratchEpochRejectsSecondExplicitContextName`; integrated `TestServeDevContextAndSwarmDirCannotPartitionScratchEpoch` |
| SCR-3 | same project/different Swarm dirs | reproduced and fixed | `TestDevScratchEpochRejectsDifferentSwarmDirOwner`; integrated context/Swarm-dir runtime test |
| SCR-4 | crash before store open | reproduced and fixed | real helper process `TestDevScratchEpochReclaimsCrashBeforeStoreOpen` |
| SCR-5 | crash after actual SQLite open | reproduced and fixed | real helper process `TestDevScratchEpochReclaimsCrashAfterStoreOpen` |
| SCR-6 | live owner mistaken as stale | reproduced and fixed | real helper process `TestDevScratchEpochDistinguishesLiveAndReleasedOwner` |
| SCR-7 | wrong project-local coordinate/normal DB mutation | reproduced and fixed | `TestDevScratchEpochUsesDedicatedProjectLocalCoordinate`; next-invocation runtime proof retains normal `dev.db` sentinel |
| SCR-8 | borrowed contracts root | reproduced and fixed | `TestDevScratchEpochRejectsBorrowedProject` with no-state-mutation assertion |
| SCR-9 | canonical project-root aliases converge; aliased internal `.swarm`/`stores` ancestors fail before mutation | reproduced and fixed | `TestDevScratchEpochCanonicalPathAliasesConverge`; `TestDevScratchEpochRejectsAliasedStateDirectoryBeforeExternalMutation`; `TestDevScratchEpochRejectsAliasedStoreDirectoryBeforeExternalMutation`; `TestDevScratchEpochCreatesMissingStateDirectories` |
| SCR-10 | custom SQLite or PostgreSQL dev | reproduced and fixed | `TestDevScratchEpochRejectsConfiguredSQLitePath`, `TestDevScratchEpochRejectsPostgres`, and supported-mode matrix |
| SCR-11 | partial DB/sidecar cleanup falls through to open | reproduced and fixed | `TestDevScratchEpochReplacementIsAllOrFailClosed` |
| SCR-12 | epoch releases before selected-store close | reproduced and fixed | mutation-sensitive `TestDevScratchEpochReleaseRequiresClosedStore`; handoff proof `TestDevScratchEpochHandsOffToSQLitePossession` |
| SCR-13 | next invocation observes predecessor facts | reproduced and fixed | `TestRunServeRuntimeDevScratchNextInvocationHasNoPredecessor`; release binary `TestGoldenAgentWorkloadSQLiteDevScratchRestartStartsFreshEpoch` with public `RUN_NOT_FOUND` readback |
| SCR-14 | dev source remains ephemeral or absent from catalog | reproduced and fixed | `TestRunServeRuntimeDevScratchPersistsExactBundleAndRunSource`; both-store `TestPrepareServeBundleSource*DevPersistsCatalogRow` |
| SCR-15 | same-invocation SQLite selected-contract fork | split / escalated as separate class | #2361, as approved by the ordered #2276 gate |
| SCR-16 | hot reload creates a second source generation | reproduced and fixed | `TestServeDoesNotExposeRetiredProjectReloadSurface` plus production constructor/hook census |

### Proof and supported surfaces

- **Whole suite at final head:** `go run ./cmd/swarm-test` - PASS.
- **Repeated lock/crash matrix:** `go test ./internal/store/devscratch -count=3` - PASS.
- **CLI admission/context grant matrix:** focused `internal/cliapp` tests - PASS.
- **Real runtime scratch/source/restart/no-reload matrix:** focused `internal/serveapp` tests - PASS.
- **Release binary supported surface:** fresh dev restart, exact health source identity, and public predecessor `run.get -> RUN_NOT_FOUND` - PASS.
- **Both selected stores:** private canonical ingest tests and supported live-source command matrix - PASS.
- **Static guards:** async work ownership ledger, persistence authority registry, public backend matrix, multi-bundle conformance matrix, and parser-backed platform spec tests - PASS.

The generic hostile proof is the helper-process kernel-lock matrix plus the close-failure mutation test. Moving acquisition after deletion, keying it by context/Swarm-dir, replacing the lock with file/PID inference, or releasing before close breaks named tests.

### Parent, watchlist, and architecture decisions

- **Sibling probe:** public SQLite registration is closed by #2359; changed-bundle loading remains #2277; DB-loaded SQLite boot remains unsupported/watchlist-only; selected-contract SQLite fork and final composition remain #2361.
- **Watchlist:** canonical node `runtime_store_backend_default_and_sqlite_portability` at `swarm-docs@18a2eee`. It already tracks crash-reclaimable coordinate authority and points to no additional live Child B consumer family. Evidence supports leaving only #2361 rather than absorbing the fork state machine here.
- **Watchlist decision:** mapping remains correct; no refinement or new issue is required.
- **Architecture smell:** the retired Builder replacement engine remains a large internal subsystem even though dev/API/test-hook reachability is gone. Long-run work should either give changed-bundle context loading a newly approved owner or delete the engine, not reconnect it opportunistically.
- **Tracking decision:** existing #2277; no new issue. Estimated effort 3-5 engineering days after its source-authority design is ruled, with medium ROI. It is non-closure-bearing for #2360.
- **Platform lock duplication:** epoch and opened-inode locks deliberately remain distinct exclusion domains. Shared low-level helpers may be considered only if a third owner appears; watchlist/residual risk only, estimated 1-2 days and low current ROI.
- **Native-platform note:** all required proofs ran on Linux. A supplemental Darwin cross-compile attempt was blocked before this package by the repository's existing `wasmtime-go` cross-build limitation; the Darwin adapter follows the same retained `flock`/single-link/filesystem contract and remains subject to normal CI/platform execution.

### Closure claim

This PR eliminates `project_local_dev_scratch_coordinate_epoch_exclusion_reclamation_replacement_and_source_ownership` entirely. Fixing only source persistence, only context claims, or only file cleanup would still leave another same-concept interpreter; after this head, none remains live. #2276 stays open only for the explicitly split #2361 child. There is no blocking ambiguity.

### Final-head rebase confirmation

- **Final audited head:** `2db582963` (clean rebase onto current `origin/master`; no conflict and no semantic diff from the audited change).
- **Required whole suite at this exact head:** `go run ./cmd/swarm-test` - PASS, including release E2E, `internal/serveapp`, `internal/store/devscratch`, runtime persistence, and PostgreSQL coverage.
- The owner census, manifestation table, split decision, and failure-class closure claim above remain unchanged and apply to this exact merge candidate. Replacement CI is running against this head.

### CI timing-repair evidence

- The first rebased CI run exposed two pre-existing timing assumptions in `internal/testpostgres`; neither failure touched the #2360 production boundary.
- `TestRunAdmissionReportsUnknownETAWhenEvidenceIsMissing` now cancels only after the queue report is observed, rather than racing ticket registration against the first report.
- `TestManagerReconcileRefreshesIntentSnapshotAfterTakingLease` now joins the account-scoped resource lease before asserting terminal cleanup when another manager wins reconciliation.
- Focused stress: both repaired tests together at `-count=50` - PASS; complete `internal/testpostgres` at `-count=3` - PASS.
- Final required whole suite after both repairs: `go run ./cmd/swarm-test` - PASS. No production code or #2360 closure accounting changed.

### Review cycle 1 repair

- **Finding closed:** pre-open state establishment no longer calls recursive `MkdirAll` through unchecked internal ancestors. `EpochAuthority` validates or creates `.swarm` and `stores` one component at a time with `Lstat` plus canonical identity, rechecks the retained project/state/store identities before descending and immediately before authority-file acquisition, and rejects aliases without adopting them.
- **Zero-mutation hostile proof:** a `.swarm` symlink to an external empty directory fails without creating external `stores` or any authority artifact; the sibling `.swarm/stores` symlink also fails with an unchanged external target.
- **First-use proof:** a project with neither directory creates exact unaliased `.swarm/stores` and acquires normally.
- **Focused evidence:** `go test ./internal/store/devscratch -count=20`; focused CLI and serve matrices at `-count=3`; release dev-scratch restart E2E - all PASS.
- **Required whole suite at final head `2db582963`:** `go run ./cmd/swarm-test` - PASS. SCR-9 and the failure-class closure claim above now include both root-alias convergence and internal-ancestor fail-closed ownership.

File: internal/store/devscratch/epoch.go [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+// predecessor files until PrepareFreshStore is called.
+func Acquire(coordinate Coordinate) (*EpochAuthority, error) {
+	canonical, err := Resolve(coordinate.ProjectRoot)
+	if err != nil {
+		return nil, err
+	}
+	if filepath.Clean(coordinate.DatabasePath) != canonical.DatabasePath || filepath.Clean(coordinate.AuthorityPath) != canonical.AuthorityPath {
+		return nil, errors.New("dev scratch coordinate does not match its canonical project owner")
+	}
+	if err := os.MkdirAll(filepath.Dir(canonical.DatabasePath), 0o700); err != nil {
```
Comment (yazzaoui) [inline internal/store/devscratch/epoch.go]: **P1 blocking:** This mutates an unverified path before the alias check below. With `<project>/.swarm` symlinked to an external empty directory, `MkdirAll` creates `<external>/stores`, then `EvalSymlinks` rejects the alias. Establish and validate the `.swarm`/`stores` components without following existing symlinks before creating anything below them, and add zero-external-mutation tests for both `.swarm` and `.swarm/stores` aliases. No compatibility adoption of aliased state is allowed.

**Review by yazzaoui — COMMENTED** [review]
**Changes needed — review cycle 1**

The issue category is correct: high-risk runtime/store/source-lifecycle architecture and conformance. The independent Gate E occurred before implementation, the exact governing `platform-spec.yaml` sections are changed in this PR, and the approved Child B boundary is honest. The chosen class is broad enough for project-local dev scratch ownership; #2361 remains the independently verified fork-parity sibling that closes parent #2276.

## Finding

**P1 blocking — internal state-directory aliases mutate a non-owned coordinate before fail-closed rejection.**

`devscratch.Acquire` calls `os.MkdirAll(<project>/.swarm/stores)` before resolving and rejecting aliases. An adversarial probe with `<project>/.swarm -> <external-empty-dir>` correctly returned the alias error, but first created `<external-empty-dir>/stores`. That falsifies the claimed fail-closed alias boundary: the epoch owner has mutated outside the canonical project before proving ownership. This is distinct from the covered project-root symlink convergence and predecessor-sidecar symlink tests.

Required one-pass repair:

1. Establish `.swarm` and `stores` component ownership without following an existing symlink. Validate each existing component with `Lstat`/canonical identity before creating the next missing component; then revalidate the final parent before opening the authority file. Do not add a compatibility path or adopt an aliased directory.
2. Add a test where `.swarm` is a symlink to an external empty directory and assert acquisition fails with zero mutation in the external target.
3. Add the sibling `.swarm/stores` symlink case and retain a normal missing-directory success case, so the fix does not accidentally reject first use.
4. Update SCR-9 in the proof audit to distinguish canonical project-root aliases, which converge, from aliased internal state ancestors, which fail before mutation.

## Independent owner/consumer/lifecycle map

- Posture/coordinate projection: `cliapp.ResolveDevScratch` admits only project-local, platform-default SQLite and rejects borrowed, custom SQLite, and PostgreSQL postures.
- Pre-open exclusion/replacement owner: `devscratch.EpochAuthority`; the adjacent retained kernel lock is the correct semantic owner.
- Open-store handoff: `devscratch.SelectedStoreLifecycle` composes the epoch with selected SQLite possession.
- Source owner: `BuildBundleCatalogProjection` plus selected `ServeIngestWriter.UpsertBundleCatalogWithData` atomically persist the exact catalog/data projection; `BundleSourceFact` is persisted.
- Descriptor consumer: context registration remains discovery metadata. The production ordering retains the epoch throughout reconciliation; it is not a second scratch-liveness interpreter.
- Shutdown: process work retires/joins, process capability settles, selected store closes, then the epoch lock releases.
- Siblings checked: normal project `dev.db`, non-dev selected SQLite/PostgreSQL, DB-loaded SQLite boot, public bundle registration, selected-contract fork, changed-bundle loading, Builder reload, context names, and alternate Swarm dirs. No missed live same-concept scratch interpreter was found.

## Adversarial matrix and closure

The code-level sweep matched the audit for concurrent owners, context/Swarm-dir independence, crash-before-open, crash-after-open, live contention, normal `dev.db` preservation, borrowed/custom/PostgreSQL refusal, sidecar validation, exact persisted source, next-epoch replacement, no production reload route, and store-close-before-unlock. It did not match SCR-9 for internal state-directory aliases, as demonstrated above.

The architecture direction is sound and materially reduces drift: one coordinate owner replaces descriptor/PID/path guesses, and one persisted source owner replaces the dev-only ephemeral branch. No broad refactor is warranted. After the narrow alias repair and exact proof update, the chosen failure class can honestly be claimed eliminated.

Tracker/watchlist state was independently verified: #2275, #2276, #2361, and #2277 are open with the recorded roles; #2359 is closed; `swarm-docs@18a2eee` contains the canonical `runtime_store_backend_default_and_sqlite_portability` mapping. No watchlist change, new issue, or `docs/POTENTIAL_ISSUES.md` entry is warranted from this pass. The retained Builder replacement subsystem is already captured by #2277.

**Comment by yazzaoui** [comment]
### Review checklist — cycle 1

- Issue/category/spec: **pass** — high-risk runtime/store/source-lifecycle work; exact authoritative spec sections updated.
- Independent pre-audit Gate E before coding: **pass**.
- Failure-class boundary and parent/sibling accounting: **pass** — Child B is class-complete in intent; #2361 remains the valid final fork sibling under #2276.
- Canonical owner and repo-wide consumer accounting: **pass**.
- Manifestation proof: **changes needed** — SCR-9 misses internal `.swarm`/`stores` aliases; independent execution proved mutation outside the project before rejection.
- Post-implementation proof audit: **changes needed** — refresh SCR-9 after code and zero-mutation proof land.
- Platform spec / no migration / no compatibility seam: **pass**.
- Watchlist/tracker: **pass** — no change; canonical mapping verified at `swarm-docs@18a2eee`; #2277 already owns retained reload architecture.
- Focused local proof: **pass except the intentional adversarial probe** — devscratch x3, CLI scratch/registration x3, serve scratch/source/no-reload x2, release restart, spec/conformance, and the two late testpostgres cases x10 all passed. The review-only alias probe failed exactly as reported and was removed.
- CI: **green** at `3372043fac4a14b312596a40b412d1bb4f7b89bc`, including required full shards and timing budget.
- Risk: **high semantic area; narrow remaining repair**.
- Closure level: **touched seam canonicalized, but failure class not yet eliminated**.
- Residual risk after repair: platform-specific lock/filesystem behavior remains bounded to the documented Darwin/Linux support posture; no additional same-concept interpreter found.
- Merge recommendation: **do not merge until the internal-ancestor alias mutation is fixed and SCR-9 proof is expanded**.

Formal `REQUEST_CHANGES` is unavailable because the authenticated reviewer is the PR author; the substantive COMMENT review is explicitly changes-needed.

File: internal/store/devscratch/epoch.go (hunk 1 above)
Comment (yazzaoui) [reply internal/store/devscratch/epoch.go]: Fixed at `2db582963`. `Acquire` now establishes `.swarm` and `stores` component-by-component: each existing component is inspected without following symlinks, canonical identity is validated, parent identities are rechecked before descending, and the complete chain is revalidated immediately before opening the authority file. Aliased components are rejected; there is no adoption path. Added zero-external-mutation proofs for both `.swarm` and `.swarm/stores`, plus explicit missing-directory first-use success. `go test ./internal/store/devscratch -count=20`, focused CLI/serve/release proofs, and `go run ./cmd/swarm-test` all pass. SCR-9 and the final-head proof audit are updated.

**Comment by yazzaoui** [comment]
### Review cycle 1 repair ready

The P1 internal-ancestor alias finding is fixed at `2db582963`. `.swarm` and `stores` are now established component-by-component with no-follow inspection/canonical identity and revalidated before authority-file acquisition. Both hostile alias cases prove zero external mutation; missing-directory first use remains supported.

The inline thread is resolved, SCR-9 and the final-head proof audit are corrected, `go run ./cmd/swarm-test` passes, and all 25 CI checks are green. Ready for independent review cycle 2.

**Review by yazzaoui — COMMENTED** [review]
**Review cycle 2 — no blocking findings; merge recommended**

Cycle 2 was required because cycle 1 independently found that stationary internal `.swarm` aliases were rejected only after `MkdirAll` mutated their external target. Commit `2db582963` changes production code, not only audit prose, and closes that finding.

## Remediation verification

- `Acquire` no longer recursively creates through unchecked ancestors.
- Existing `.swarm` and `stores` components are inspected without following a leaf symlink, canonical identity is checked, and identities are rechecked before authority-file acquisition.
- Missing first-use directories are created one component at a time.
- The requested hostile cases now have separate execution proofs: aliased `.swarm`, aliased `.swarm/stores`, and ordinary missing-directory creation.
- SCR-9 now distinguishes canonical project-root aliases, which converge, from internal state ancestors, which fail before mutation.
- The prior inline finding is resolved and obsolete on the current head.

I reran the static alias scenarios and the complete focused Child B matrix. The original external-mutation reproducer no longer manifests.

## Architecture reassessment at cycle 2

The deeper design remains sound: `ResolveDevScratch` owns posture/coordinate admission; `EpochAuthority` owns pre-open exclusion and replacement; `SelectedStoreLifecycle` composes opened SQLite possession with close-before-unlock; selected serve ingest owns exact persisted source; context descriptors remain discovery metadata only. No second live scratch/source interpreter was found.

I also adversarially tested a different boundary: a same-user process can rename the project-owned `stores` directory and replace it with a symlink *after* valid acquisition; path-based replacement can then follow the new namespace. The repository's existing SQLite possession owner has the same local-filesystem trust posture. This is not a supported Swarm mutation or a privilege boundary: that same principal can already directly rewrite/delete the selected database. Closing it absolutely would require a substantially broader cross-platform retained-dirfd plus SQLite construction/VFS design, not another local recheck. Under the explicit no-over-engineering instruction and the approved #2360 class, I classify this as out-of-model residual risk, not a missed same-concept consumer or merge blocker. No tracker promotion is warranted absent a product requirement to defend runtime state from the owning OS user.

## Closure assessment

- Issue category, independent Gate E, exact spec updates: pass.
- Chosen Child B class: broad enough and eliminated.
- Parent/sibling state: valid; #2361 remains the independently verified final fork-parity child of #2276.
- Canonical owner and repo-wide consumer census: pass.
- All currently known in-model manifestations: explicitly proven.
- Concept-wide drift reduction: material; descriptor/PID/path guesses and dev-only ephemeral source interpretation are gone.
- Same-concept interpreters remaining: none.
- Closure level: failure class eliminated.
- Watchlist: `runtime_store_backend_default_and_sqlite_portability` at `swarm-docs@18a2eee` remains correct; no change.
- Architecture follow-up: retained Builder replacement machinery remains already captured by #2277; no new issue or `docs/POTENTIAL_ISSUES.md` entry.

Focused local proof on `2db582963978b61e0ee1105ff74bab8fb6c32069`:

- `go test ./internal/store/devscratch -count=20`
- focused `internal/cliapp` scratch/registration matrix at `-count=3`
- focused `internal/serveapp` scratch/source/no-reload matrix at `-count=3`
- release restart E2E
- API spec and conformance projections

All passed. All required CI jobs, including the full shards and timing budget, are green at this exact head. Merge recommendation: **merge after the repository's independent-approval requirement is satisfied**. Formal APPROVE is unavailable because the authenticated reviewer owns the PR.

**Comment by yazzaoui** [comment]
### Review checklist — cycle 2

- Prior P1 implementation repair: **verified closed on current head**.
- Issue/category/exact spec: **pass**.
- Independent Gate E before coding: **pass**.
- Failure-class and parent/sibling accounting: **pass**.
- Canonical owner / repo-wide consumer census: **pass**.
- Manifestation table and exact proof: **pass**, including repaired SCR-9.
- No legacy, migration, compatibility, or dual-owner seam: **pass**.
- Concept-wide semantic drift reduction: **pass**.
- Remaining live same-concept interpreter: **none found**.
- Post-implementation proof audit: **pass** at `2db582963`.
- Focused local tests: **pass**.
- Exact-head required CI: **green**.
- Risk: **high semantic area, implementation proof complete**.
- Closure level: **failure class eliminated**.
- Residual risk: same-user post-acquisition filesystem namespace replacement is outside the existing local-store trust boundary; absolute defense would require a broader dirfd/SQLite-VFS architecture and is not justified for this class.
- Watchlist action: **no change**; existing mapping verified.
- Follow-up: **#2277 already captures retained source-replacement architecture; #2361 remains the valid parent sibling**.
- Merge recommendation: **merge after independent approval requirement is satisfied**.

Cycle 2 reason: cycle 1 exposed a stationary internal-alias mutation absent from the original proof. The production repair and three exact tests now close it. Architecture reassessment found no deeper in-scope ownership problem and no reason for another revision cycle. Formal `APPROVE` is unavailable because the authenticated reviewer owns the PR.
